src/formula/variant.cpp

```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
+}
 
-	n -= decimal_value_;
+variant::variant(int n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
+}
 
-	if(n > 0.5)
-		decimal_value_++;
-	else if(n < -0.5)
-		decimal_value_--;
+variant::variant(double n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
 }
 
 variant::variant(const game_logic::formula_callable* callable)
-	: type_(TYPE_CALLABLE), callable_(callable)
+	: value_(std::make_shared<game_logic::variant_callable>(callable))
 {
-	assert(callable_);
+	assert(value_.get());
 }
 
-variant::variant(std::vector<variant>* array)
-    : type_(TYPE_LIST)
+variant::variant(std::vector<variant>* vec)
+    : value_((std::make_shared<game_logic::variant_list>(*vec)))
 {
-	assert(array);
-	list_ = new std::vector<variant>(*array);
+	assert(value_.get());
 }
 
 variant::variant(const std::string& str)
-	: type_(TYPE_STRING)
+	: value_(std::make_shared<game_logic::variant_string>(str))
 {
-	string_ = new std::string(str);
+	assert(value_.get());
 }
 
 variant::variant(std::map<variant,variant>* map)
-    : type_(TYPE_MAP)
+	: value_((std::make_shared<game_logic::variant_map>(*map)))
 {
-	assert(map);
-	map_ = new std::map<variant,variant>(*map);
+	assert(value_.get());
 }
 
 variant::variant(const variant& v)
-    : type_(v.type_)
-{
-	switch(type_) {
-		case TYPE_INT:
-			int_value_ = v.int_value_;
-			break;
-		case TYPE_DECIMAL:
-			decimal_value_ = v.decimal_value_;
-			break;
-		case TYPE_LIST:
-			list_ = new std::vector<variant>(*v.list_);
-			break;
-		case TYPE_STRING:
-			string_ = new std::string(*v.string_);
-			break;
-		case TYPE_MAP:
-			map_ = new std::map<variant,variant>(*v.map_);
-			break;
-		case TYPE_CALLABLE:
-			callable_ = v.callable_;
-			break;
-		case TYPE_NULL:
-			break;
-	}
-}
-
-variant::~variant()
+    : value_(v.value_)
 {
-	release();
 }
 
 variant& variant::operator=(const variant& v)
 {
-	if(&v != this) {
-		this->~variant();
-		new(this) variant(v);
-	}
+	value_ = v.value_;
 	return *this;
 }
 
 variant variant::operator[](size_t n) const
 {
-	if(type_ == TYPE_CALLABLE) {
+	if(is_callable()) {
 		return *this;
 	}
 
-	must_be(TYPE_LIST);
-	assert(list_);
-	if(n >= list_->size()) {
+	must_be(VARIANT_TYPE::TYPE_LIST);
+
+	try {
+		return value_cast<game_logic::variant_list>()->get_container()[n];
+	} catch(std::out_of_range&) {
 		throw type_error("invalid index");
 	}
-
-	return (*list_)[n];
 }
 
 variant variant::operator[](const variant& v) const
 {
-	if(type_ == TYPE_CALLABLE) {
+	if(is_callable()) {
 		return *this;
 	}
 
-	if(type_ == TYPE_MAP) {
-		assert(map_);
-		std::map<variant,variant>::const_iterator i = map_->find(v);
-		if (i == map_->end())
-		{
+	if(is_map()) {
+		auto& map = value_cast<game_logic::variant_map>()->get_container();
+
+		auto i = map.find(v);
+		if(i == map.end()) {
 			static variant null_variant;
 			return null_variant;
```

Comment:
Well, nothing wrong with this, but I wonder why it bothered. It could just `return variant()` for the same effect, and it's not like `variant` is a heavyweight class that's expensive to construct.